logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Listener.java

```diff
@@ -36,7 +35,12 @@ public void onWrite(final long address) {
 
   @Override
   public void onWriteError(final Throwable error) {
-    if (error instanceof NoSuchElementException || error instanceof NoLeader) {
+    if (error instanceof NoSuchElementException
+        || error
+            .getMessage()
+            .contains(
+                "Failed to locate leader") // this is sub-optimal and should be checked elsewhere
+    ) {
       // Not a failure. It is probably during transition to follower.
```

Comment:
:wrench: I would like to remove this

Reply:
I agree that this should be hidden, as it leaks knowledge of Atomix. This should be pushed down either to the LogStorage or to the ZeebeLogAppender (or whatever the implementation is called :smile:)

Reply:
Do we need it at all? If we anyway not the leader we do not care right? :shrug:

Reply:
If the write fails due to other error, we would like to catch it and stop LogStorageAppender to stop writing new records.

Reply:
Exactly, so why we make here the difference with the exception? We can handle both the same or not?

Reply:
If it is not leader, then it is an expected case because of asynchronous leader transition. We can just ignore them - as we do here. We just log it in debug and continue. The other error cause it to go unhealthy, if I remember correctly.

Reply:
Can we not handle it on our side and simply return a `NoSuchElementException` or a new exception that is part of the API, and not an Atomix detail?

Reply:
That would be possible.

Reply:
I feel we talk past each other. I see the point that we CAN ignore it in the cause, but it is not necessary, right? Means we could just handle it the same which would simplify our code here and reduce the dependencies. Currently, the code has knowledge of atomix internals which is sub-optimal. 

@npepinpe I haven't looked into much detail here, but I can do if you want. I first wanted to discuss this gem. :gem:

Reply:
> I feel we talk past each other. I see the point that we CAN ignore it in the cause, but it is not necessary, right? Means we could just handle it the same which would simplify our code here and reduce the dependencies. Currently, the code has knowledge of atomix internals which is sub-optimal.

If you handle it the same, there will be error logs whenever a leader transition to follower. One error which says "Failed to append block .." and the second error which logs this partition as unhealthy.

Reply:
I will take a look how I could circle around it

Reply:
@deepthidevaki and @npepinpe what do you think of the last commit https://github.com/camunda/zeebe/pull/9235/commits/ae15d717561119435fd546d4a5b8e536cc2e2cde I moved the handling into the adapter. BTW for what is the `NoSuchElementException` handling I found no place where this can be caused?